lib/navigation/preprocess_encoders.cpp

```diff
@@ -2,55 +2,99 @@
 {
-  // TODOLater: implement
 }
 
-core::EncoderData EncodersPreprocessor::processData(const core::EncoderData)
+std::optional<core::EncoderData> EncodersPreprocessor::processData(
+  const core::EncoderData raw_encoder_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkRelaible
-
-  */
-  return {0, 0, 0, 0};
+  const auto encoder_data = detectOutliers(raw_encoder_data);
+  SensorChecks sensors    = checkReliable();
+  if (!encoder_data.has_value() || sensors == SensorChecks::kUnacceptable) {
+    return std::nullopt;
+  } else {
+    return encoder_data;
+  }
 }
 
-core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data)
+std::optional<core::EncoderData> EncodersPreprocessor::detectOutliers(
+  const core::EncoderData encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::EncoderData encoder_data_copy;
+  std::copy(encoder_data.begin(), encoder_data.end(), encoder_data_copy.begin());
+  Quartile quartiles;
+  core::Float interquartile_range;
+  core::Float upper_bound;
+  core::Float lower_bound;
+  // Case 1: All sensors are functioning properly
+  if (num_reliable_encoders_ == core::kNumEncoders) {
+    core::EncoderData reliable_data;
+    std::copy(encoder_data.begin(), encoder_data.end(), reliable_data.begin());
+    quartiles           = getQuartiles(reliable_data);
+    interquartile_range = quartiles.q3 - quartiles.q1;
+    upper_bound         = quartiles.median + 1.5 * interquartile_range;
+    lower_bound         = quartiles.median - 1.5 * interquartile_range;
+
+  }
+  // Case 2: One of the sensors has become faulty
+  else if (num_reliable_encoders_ == core::kNumEncoders - 1) {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t counter = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (encoders_reliable_.at(i)) {
+        reliable_data.at(counter) = encoder_data.at(i);
+        ++counter;
+      }
+    }
+    quartiles           = getQuartiles(reliable_data);
+    interquartile_range = quartiles.q3 - quartiles.q1;
+    upper_bound         = quartiles.median + 1.2 * interquartile_range;
+    lower_bound         = quartiles.median - 1.2 * interquartile_range;
+  }
+  // Case 3: More than one sensors have become unreliable, stopping everything and entering into
+  // fail State.
+  else {
+    logger_.log(core::LogLevel::kFatal,
+                "Number of unreliable encoder sensors have exceeded the threshold");
+    return std::nullopt;
+  }
```

Comment:
i would move this to a guard clause before the if statements, then `else if` on line 45 can just be `if`

Reply:
No because it directly assigns values to an instantiated object depending on the conditions. Surely this is exactly the use case for `if`, `else if`, `else`?

Reply:
idk about `if`, `else if` but there is definitely no need to set up all the variables and copy the encoder data around if you can already determine `num_reliable_encoders_ < core::kNumEncoders - 1` as soon as the function is called

Reply:
I mean we could check this closer to the start but the reason we put it here was to give a default case of failure in case of bugs, etc. This is just an additional precaution to avoid undefined behavior in the logic of this function (we should never reach this if statement but its good to have it surely?). As for the copies, this is because the thresholds should be different depending on if we have unreliable sensors already and the readings need to be stored in different sized arrays since we don't want to be using anything that does not have a statically defined size.

Reply:
it makes sense to have, but it should be the first thing in the function. i understand why the copy of the data is made, but in the case it is already provable at the start of the function that `num_reliable_encoders_ < core::kNumEncoders - 1` and therefore the function should exit there. making the copy means allocating memory in a case where we know it is going to do nothing but be deallocated immediately after, which is pointless

Reply:
The container which we copy the array to is also an array with a statically defined size. Therefore the space is already allocated to stack memory so we're not actually allocating memory, just reassigning the values stored at an address (I think? @Fayz-S?). I fully see what you mean but also I feel like it's just removing a safeguard in the form of a default fail state?